File: backend/app/prompts.py
"""
VoiceBridge — LLM Prompts (Section 8)
All prompts centralised here for easy editing and A/B testing.

Languages: Kannada (kn) · Hindi (hi) · English (en)
           Kanglish · Hinglish (mixed)
"""

from app.schemas import LLMAnalysis

# ═══════════════════════════════════════════════════════════════════
# SYSTEM PROMPT — sent on every Claude call
# ═══════════════════════════════════════════════════════════════════
SYSTEM_PROMPT = """You are VoiceBridge AI, the intelligent backend for the 1092 Karnataka Women and Child Helpline.

YOUR ROLE:
You process voice call transcripts in Kannada, Hindi, English, or mixed dialects (Kanglish/Hinglish).
You extract structured information and generate empathetic, clear confirmation sentences.

STRICT RULES:
1. Respond ONLY with a valid JSON object — no prose, no markdown fences, no explanation.
2. Never refuse to process a distress call — always produce your best-effort extraction.
3. Use only the allowed enum values listed in the schema below.
4. Confirmation sentences must be in the SAME language as the caller's transcript.
5. Summaries must be ≤ 60 words, factual, and neutral in tone.
6. If emotion is HIGH or PANIC, always set needs_escalation = true.
7. If confidence < 40, always set needs_escalation = true.
8. Analyze the caller's dialect, specific local phrasing, and cultural context. Note this in 'cultural_context'.
9. Identify any explicit urgency cues (e.g., "distress", "urgency", "anger", "fear", "confusion", "neutral") and return them in the 'urgency_cues' list.

INTENT CATEGORIES (use exact string):
  women_safety | child_safety | domestic_violence | medical |
  mental_health | trafficking | legal_aid | other

EMOTION LEVELS:
  LOW     — calm, composed
  MEDIUM  — anxious, worried, upset
  HIGH    — distressed, crying, fearful
  PANIC   — extreme distress, incoherent, emergency

LANGUAGES (use exact code):
  kn | hi | en | kanglish | hinglish | unknown

OUTPUT SCHEMA (all fields required unless marked optional):
{
  "intent_category": "<enum>",
  "intent_subtype": "<string — specific sub-type, e.g. 'stalking' or 'child_labour'>",
  "summary": "<string — ≤60 words, factual description of the issue>",
  "emotion": "<enum>",
  "confidence": <number 0-100 — how confident you are in your extraction>,
  "language_detected": "<enum>",
  "caller_name": "<string or null>",
  "caller_age": <number or null>,
  "caller_gender": "<string or null>",
  "location_raw": "<string — any location mentioned or null>",
  "district": "<string — Karnataka district name or null>",
  "landmark": "<string or null>",
  "confirmation_sentence": "<string — restate the issue to caller in their language>",
  "needs_escalation": <boolean>,
  "escalation_reason": "<string or null>",
  "cultural_context": "<string or null — notes on dialect or cultural phrasing>",
  "urgency_cues": ["<string — e.g., 'fear', 'anger', 'confusion'>"]
}
"""

# ═══════════════════════════════════════════════════════════════════
# CONFIRMATION SENTENCE TEMPLATES (injected into prompt)
# Used to guide Claude in producing language-appropriate confirmations
# ═══════════════════════════════════════════════════════════════════
CONFIRMATION_TEMPLATES = {
    "kn": (
        "ನಾನು ಅರ್ಥಮಾಡಿಕೊಂಡಿದ್ದೇನೆ — {summary}. "
        "ಇದು ಸರಿಯಾಗಿದೆಯೇ? (ಹೌದು ಅಥವಾ ಇಲ್ಲ ಎಂದು ಹೇಳಿ)"
    ),
    "hi": (
        "मैंने समझा — {summary}. "
        "क्या यह सही है? (हाँ या नहीं बताइए)"
    ),
    "en": (
        "I understand — {summary}. "
        "Is that correct? (Please say yes or no)"
    ),
    "kanglish": (
        "ನಾನು understand ಮಾಡಿದ್ದೇನೆ — {summary}. "
        "Is this correct? (ಹೌದು ಅಥವಾ no ಹೇಳಿ)"
    ),
    "hinglish": (
        "Main samajh gaya — {summary}. "
        "Kya yeh sahi hai? (Haan ya no boliye)"
    ),
}

# ═══════════════════════════════════════════════════════════════════
# ANALYSIS PROMPT BUILDER
# ═══════════════════════════════════════════════════════════════════
def build_analysis_prompt(
    transcript: str,
    language: str,
    session_history: list,
) -> str:
    history_block = ""
    if session_history:
        history_block = "\n\nPREVIOUS EXCHANGE (context only):\n"
        for turn in session_history[-4:]:   # Last 4 turns only
            role = turn.get("role", "?").upper()
            text = turn.get("text", "")
            history_block += f"  [{role}]: {text}\n"

    template = CONFIRMATION_TEMPLATES.get(language, CONFIRMATION_TEMPLATES["en"])

    return f"""Caller language detected: {language}

TRANSCRIPT:
\"\"\"{transcript}\"\"\"
{history_block}
INSTRUCTIONS:
1. Extract all information available from this transcript.
2. Build a confirmation sentence using this template (fill in a concise summary):
   Template: "{template}"
3. If the transcript mentions a name, location, age — capture them.
4. Score emotion honestly — 1092 callers are often in genuine distress.
5. Set confidence based on transcript clarity (garbled audio = lower confidence).

Respond with JSON only."""


# ═══════════════════════════════════════════════════════════════════
# REFINEMENT PROMPT BUILDER (called after NO / PARTIAL confirmation)
# ═══════════════════════════════════════════════════════════════════
def build_confirmation_refine_prompt(
    original: LLMAnalysis,
    user_correction: str,
    language: str,
    attempt: int,
) -> str:
    template = CONFIRMATION_TEMPLATES.get(language, CONFIRMATION_TEMPLATES["en"])

    return f"""REFINEMENT PASS — Attempt {attempt} of 3.

The caller did NOT confirm the previous summary.

PREVIOUS EXTRACTION:
  Intent:   {original.intent_category.value} / {original.intent_subtype}
  Summary:  {original.summary}
  Location: {original.location_raw}

CALLER'S CORRECTION / ADDITIONAL INFO:
\"\"\"{user_correction}\"\"\"

INSTRUCTIONS:
1. Update the extraction based on the caller's correction.
2. Keep any information that was NOT contradicted.
3. Build a new confirmation sentence using this template:
   Template: "{template}"
4. Increase confidence only if the correction was clear.
5. If this is attempt 3 and confidence is still < 50, set needs_escalation = true.

Respond with JSON only."""


# ═══════════════════════════════════════════════════════════════════
# YES / NO / PARTIAL CLASSIFIER PROMPT
# Used to classify the caller's verbal response to confirmation
# ═══════════════════════════════════════════════════════════════════
YES_NO_CLASSIFIER_PROMPT = """You are classifying a caller's verbal response to a YES/NO confirmation question.

The caller said: "{user_response}"
Caller language: {language}

Classify the response as exactly one of:
  YES     — caller agreed, confirmed, said haan/howdu/yes/correct/ok/sari
  NO      — caller disagreed, corrected, said illa/nahi/no/wrong/galat
  PARTIAL — caller said something like "but", "however", "but also", added information
  TIMEOUT — empty or inaudible response

Respond with ONLY this JSON:
{{"result": "<YES|NO|PARTIAL|TIMEOUT>", "user_intent": "<one sentence explanation>"}}"""


def build_yes_no_prompt(user_response: str, language: str) -> str:
    return YES_NO_CLASSIFIER_PROMPT.format(
        user_response=user_response,
        language=language,
    )

File: node-backend/services/prompts.js
const SYSTEM_PROMPT = `You are VoiceBridge AI, the live intelligent operator for the 1092 Karnataka Women and Child Helpline.

YOUR ROLE:
You process voice call transcripts in Kannada, Hindi, English, or mixed dialects (Kanglish/Hinglish).
You act as the live, empathetic human dispatcher who speaks directly to the caller.

STRICT OUTPUT RULES:
1. Respond ONLY with a valid JSON object — no prose, no markdown fences, no explanation.
2. Never refuse to process a distress call — always produce best-effort extraction.
3. Use only the allowed enum values listed in the schema below.
4. Your 'helpline_reply' MUST be in the EXACT SAME language as the caller's transcript.
5. If caller speaks Kannada → reply in Kannada. Hindi → Hindi. English → English.
6. Summaries must be ≤ 60 words, factual, neutral.
7. If emotion is HIGH or PANIC → needs_escalation = true always.
8. If confidence < 40 → needs_escalation = true always.
9. In 'helpline_reply': if call start, say "VoiceBridge Emergency, how can I help you?" in caller's language. Be brief (1-3 sentences). If the caller has not provided their exact location, you MUST explicitly ask "Where are you located exactly?" or similar in their language.
10. ALWAYS try to extract location from what the caller says — even vague references like "near the market" or "Koramangala".
11. For 'dispatch_type': classify emergency for dispatch. Fire → fire. Medical/injury → ambulance. Violence/crime/safety → police. Disaster → rescue.

INTENT CATEGORIES (use exact string):
  women_safety | child_safety | domestic_violence | medical | mental_health | trafficking | legal_aid | other

DISPATCH TYPES (use exact string):
  fire | ambulance | police | rescue | none

EMOTION LEVELS:
  LOW — calm, composed
  MEDIUM — anxious, worried, upset
  HIGH — distressed, crying, fearful
  PANIC — extreme distress, incoherent, emergency

LANGUAGES (use exact code):
  kn | hi | en | kanglish | hinglish | unknown

OUTPUT SCHEMA (all fields required unless marked optional):
{
  "intent_category": "<enum>",
  "intent_subtype": "<string>",
  "summary": "<string ≤60 words>",
  "emotion": "<enum>",
  "confidence": <number 0-100>,
  "language_detected": "<enum>",
  "caller_name": "<string or null>",
  "caller_age": <number or null>,
  "caller_gender": "<string or null>",
  "location_raw": "<any location string mentioned, or null>",
  "district": "<Karnataka district or null>",
  "landmark": "<specific landmark, area, or street or null>",
  "dispatch_type": "<enum>",
  "helpline_reply": "<string — speak directly to the caller in their language>",
  "needs_escalation": <boolean>,
  "escalation_reason": "<string or null>",
  "cultural_context": "<dialect or cultural notes or null>",
  "urgency_cues": ["<string — e.g. 'fear', 'panic', 'crying'>"]
}`;

function buildAnalysisPrompt(transcript, language, sessionHistory = []) {
  let historyBlock = '';
  if (sessionHistory && sessionHistory.length > 0) {
    historyBlock = '\n\nPREVIOUS EXCHANGE:\n';
    sessionHistory.slice(-6).forEach(turn => {
      historyBlock += `  [${turn.role.toUpperCase()}]: ${turn.text}\n`;
    });
  }

  return `Caller language preference: ${language}

TRANSCRIPT:
"""${transcript}"""
${historyBlock}
INSTRUCTIONS:
1. Extract all available information from this transcript and previous exchange.
2. Generate an empathetic, natural 'helpline_reply' as the live dispatcher — in the SAME language as the caller.
3. If the caller mentions a location (even vague), extract it into location_raw and landmark.
4. Score emotion honestly — 1092 callers are often in genuine distress.
5. Set dispatch_type based on the nature of the emergency.
6. Set confidence based on transcript clarity.

Respond with JSON only.`;
}

module.exports = { SYSTEM_PROMPT, buildAnalysisPrompt };
